torch/_inductor/codegen/cpp.py

```diff
@@ -3903,6 +3904,128 @@ def can_fuse_vertical(self, node1, node2):
+                        isinstance(split_number, sympy.core.numbers.Integer)
+                        and isinstance(split_var, sympy.core.symbol.Symbol)
+                        and divide_index_name is not None
+                    )
+                return any(has_div(arg, name) for arg in expr.args)
+
+            return (
+                len(
+                    [
+                        expr
+                        for name, expr in node._body.indexing_exprs.items()
+                        if has_div(expr, name)
+                    ]
+                )
+                == 1
+            )
+
+        # Only one node contains a division, and the split dimension is contiguous in all other indexing_exprs.
+        nodes_have_div = [node for node in nodes if has_one_div(node)]
+        if len(nodes_have_div) != 1 or any(
+            stride_at_vec_range(
+                expr, split_var, cpu_vec_isa.pick_vec_isa().nelements(torch.float32)
+            )
+            != 1
+            for name, expr in nodes_have_div[0]._body.indexing_exprs.items()
+            if name != divide_index_name
+        ):
+            return nodes
+
+        for node in nodes:
+            _, original_body, _ = node.node.get_default_sizes_body()
+            current_iter_ranges = [*node._body.var_ranges.values()]
+            current_iter_vars = [*node._body.var_ranges.keys()]
+            split_idx = current_iter_vars.index(split_var)
+            assert split_idx < len(current_iter_vars)
+            iter_ranges: Tuple[sympy.Expr, ...] = ()
+            for idx in range(len(current_iter_ranges)):
+                if idx != split_idx:
+                    iter_ranges += (current_iter_ranges[idx],)
+                else:
+                    val_to_split = current_iter_ranges[idx]
+                    iter_ranges += (val_to_split // split_number, split_number)
+            reduce_ranges: Tuple[sympy.Expr, ...] = ()
+            (iter_vars, reduce_vars), var_ranges = dependencies.index_vars_no_squeeze(
+                iter_ranges, reduce_ranges, prefix="z"
+            )
+
+            def new_indexing_from_args(indices):
+                new_index = sympy.core.symbol.Symbol(
+                    "z" + str(len([*node._body.var_ranges.values()])),
+                    integer=True,
+                    nonnegative=True,
+                )
+                replacements = {split_var: split_var * split_number + new_index}  # type: ignore[operator]
+                return {
+                    name: sympy_subs(expr, replacements)
+                    for name, expr in node._body.indexing_exprs.items()
+                }
+
+            # Here decide the final loop order
+            reduce_vars = []
+            with patch.object(
+                original_body, "indexing_from_args", new_indexing_from_args
+            ):
+                node._body = ir.LoopBody(
+                    original_body, [iter_vars, reduce_vars], var_ranges
+                )
```

Comment:
Can we just re-tracing the loop body with the new iteration variables/expressions?

Reply:
Done.